### PR TITLE
Datasource exchange

### PR DESCRIPTION
**Pull request by PietroPasotti**
[`grafana_datasource_exchange`](https://github.com/canonical/charm-relation-interfaces/pull/207) reference implementation library

Rebase on  main after #108 is merged

### CODE REVIEW

File: src/cosl/interfaces/datasource_exchange.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    GrafanaDatasource,
+    GrafanaSourceAppData,
+)
+from ops import CharmBase
+from typing_extensions import TypedDict
+
+import cosl.interfaces.utils
+from cosl.interfaces.utils import DataValidationError
+
+log = logging.getLogger("_cluster")
```
Comment (michaeldmitry) [inline src/cosl/interfaces/datasource_exchange.py:43]: wrong module name

File: src/cosl/interfaces/datasource_exchange.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        all_relations = chain(
+            charm.model.relations[provider_endpoint], charm.model.relations[requirer_endpoint]
+        )
+
+        # filter out some common unhappy relation states
+        self._relations: List[ops.Relation] = [
+            rel for rel in all_relations if (rel.app and rel.data)
+        ]
+
+    def submit(self, raw_datasources: Iterable[DatasourceDict]):
```
Comment (michaeldmitry) [inline src/cosl/interfaces/datasource_exchange.py:107]: `submit` sounds weird to me :D can it be `publish_datasources` or even just `publish`

File: src/cosl/interfaces/datasource_exchange.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        """Submit these datasources to all remotes.
+
+        This operation is leader-only.
+        """
+        # sort by UID to prevent endless relation-changed cascades if this keeps flapping
+        encoded_datasources = json.dumps(sorted(raw_datasources, key=lambda raw_ds: raw_ds["uid"]))
```
Comment (michaeldmitry) [inline src/cosl/interfaces/datasource_exchange.py:113]: should we add validations to check for the `type` if its a valid value?

**Review by michaeldmitry — COMMENTED** [review]
Overall logic looks good to me :) few minor comments

File: src/cosl/interfaces/datasource_exchange.py (hunk 3 above)
Comment (PietroPasotti) [reply src/cosl/interfaces/datasource_exchange.py:113]: I thought about that, but the upstream docs don't have a clear enumeration of what values are valid. We only know what the builtin types are, but users can implement their own datasource plugins to extend the list. So we can't be strict either way.

File: src/cosl/interfaces/datasource_exchange.py (hunk 2 above)
Comment (PietroPasotti) [reply src/cosl/interfaces/datasource_exchange.py:107]: mm, don't mind 'publish'.

File: src/cosl/interfaces/datasource_exchange.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""Shared utilities for the inter-coordinator "grafana_datasource_exchange" interface.
+
+See https://github.com/canonical/charm-relation-interfaces/pull/207 for the interface specification.
```
Comment (michaeldmitry) [inline src/cosl/interfaces/datasource_exchange.py:7]: here as well

File: tests/test_datasource_exchange.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                self, provider_endpoint=prov, requirer_endpoint=req
+            )
+
+    # WHEN any event is processed
+    ctx = Context(BadCharm, meta={"name": "bob", **meta})
+    ctx.run(ctx.on.update_status(), State())
+    # THEN no exception is raised
+
+
+def test_ds_submit():
```
Comment (michaeldmitry) [inline tests/test_datasource_exchange.py:99]: nit: its `publish` now

**Review by michaeldmitry — APPROVED** [review]

File: pyproject.toml [hunk 6]
```diff
@@ -6,19 +6,21 @@ build-backend = "hatchling.build"
 name = "cosl"
 version = "0.0.44"
 authors = [
-  { name="sed-i", email="82407168+sed-i@users.noreply.github.com" },
+    { name = "sed-i", email = "82407168+sed-i@users.noreply.github.com" },
```
Comment (michaeldmitry) [inline pyproject.toml:9]: needs a vbump

File: pyproject.toml [hunk 7]
```diff
@@ -6,19 +6,21 @@ build-backend = "hatchling.build"
...
-  "typing-extensions",
-  "lightkube>=v0.15.4"
+    "ops",
+    "pydantic",
+    "tenacity",
+    "jsonschema",
+    "PyYAML",
+    "typing-extensions",
+    "lightkube>=v0.15.4",
+    "charm-relation-interfaces @ git+https://github.com/canonical/charm-relation-interfaces",
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply pyproject.toml:23]: it already is on main; charm-relation-interfaces is not on pypi so we need to use it as a git dependency

File: pyproject.toml [hunk 8]
```diff
@@ -99,3 +102,8 @@ reportTypeCommentUsage = false
 [tool.codespell]
 skip = ".git,.tox,build,lib,venv*,.mypy_cache"
 ignore-words-list = "assertIn"
+
+[tool.hatch.metadata]
+# allow git+ dependencies in pyproject
+allow-direct-references = true
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply pyproject.toml:108]: as above

**Pull request merged**